### ISSUE TITLE
Add support for delayed / deferred responses

### ISSUE THREAD

**Issue by tillrohrmann** [restatedev/sdk-typescript#183]
One of the missing building blocks for the Typescript SDK is to be able to respond to a call not immediately but at a later point. Currently, this is not directly possible. Instead, users need to employ the following pattern: 

1. Creating an `Awakeable` at the call site
2. Sending the `AwakeableId` via a background call to the callee
3. The callee stores the `AwakeableId` in state and can respond at a later point

```
const driverPromise = ctx.awakeable<string>();
ctx.send(driverpool.service).requestDriverForOrder(DEMO_REGION, { promiseId: driverPromise.id });
const driverId = await driverPromise.promise;
```

It would be nicer if one could defer the response w/o the caller knowing about it. One idea could be to surface the caller id and being able to terminate an invocation w/o returning a response:

```
async function assignDelivery(ctx: restate.RpcContext, _region: string, driverId: DriverId): Promise<DeliveryRequest> {
  // if we have deliveries already waiting, assign those
  const pendingDeliveries = await ctx.get<PendingDelivery[]>(PENDING_DELIVERIES);

  if (pendingDeliveries && pendingDeliveries.length > 0) {
    const deliveryRequest = assignClosestDelivery(ctx, driverId, pendingDeliveries);
    ctx.sendResponse(deliveryRequest.callerId, driverId);
    return deliveryRequest.deliveryRequest;
  }

  const availableDrivers = await ctx.get<PendingDriverRequest[]>(AVAILABLE_DRIVERS_STATE) ?? [];
  availableDrivers.push({ driverId: driverId, callerId: ctx.callerId });
  ctx.set(AVAILABLE_DRIVERS_STATE, availableDrivers);

  // defer response to later
  ctx.deferResponse();
}
```

**Comment by slinkydeveloper** [comment]
The API proposal is pretty much the following:

```
// A opaque string the user can keep around/store in state/send somewhere, to be used to send back the respone
ctx.callerId
```

```
// This is basically a trap that stops the current handler (very much like suspensions), closing the lifecycle of the invocation, without sending back a response to the caller.
// After ctx.deferResponse() is invoked, the service key is free-ed, so another request can be executed, while the caller will remain waiting on a response.
ctx.deferResponse()
```

```
// callerId is the aforementioned opaque string
// T is the response.
// One can potentially invoke this method multiple times with different callerId(s).
// This method is idempotent. Once sent the response once, sending it again to the same callerId won't have any effect.
ctx.sendResponse<T>(callerId, T);
```

**Comment by tillrohrmann** [comment]
An alternative idea for the API could be returning some form of promise that indicates that the response will be sent later.

**Comment by slinkydeveloper** [comment]
> An alternative idea for the API could be returning some form of promise that indicates that the response will be sent later.

What needs to be understood about that is how would the interaction with external clients work

**Comment by tillrohrmann** [comment]
> What needs to be understood about that is how would the interaction with external clients work. I guess you send back some id to pass through a restate defined endpoint to poll the result back?

Returning a promise from the function is only a way to terminate the function invocation (kind of syntactic sugar). Under the hood, nothing would be sent back to the client, I believe.

**Comment by slinkydeveloper** [comment]
I guess it can work out, but only in javascript/typescript. In other strongly typed languages we probably still need the trap method.

**Comment by slinkydeveloper** [comment]
Nothing actionable right now
